tests/types.test.ts

```diff
@@ -0,0 +1,273 @@
+      validator({
+        choices: ['foo', 'bar'],
+        default: 'foo',
+      }),
+    ).toEqualTypeOf<RequiredValidatorSpec<'foo' | 'bar'>>()
+    //@ts-expect-error - baz is not assignable to 'foo' | 'bar'
+    validator({ choices: ['foo', 'bar'], default: 'baz' })
+    // Basic
+    expectTypeOf(validator()).toEqualTypeOf<RequiredValidatorSpec<string>>()
+    expectTypeOf(validator<'foo'>()).toEqualTypeOf<RequiredValidatorSpec<'foo'>>()
+    expectTypeOf(validator({ default: undefined })).toEqualTypeOf<OptionalValidatorSpec<string>>()
+    expectTypeOf(validator({ devDefault: undefined })).toEqualTypeOf<
+      RequiredValidatorSpec<string>
+    >()
+
+    expectTypeOf(validator({ default: undefined })).toEqualTypeOf<OptionalValidatorSpec<string>>()
+    expectTypeOf(validator({ devDefault: undefined })).toEqualTypeOf<
+      RequiredValidatorSpec<string>
+    >()
+
+    expectTypeOf(validator({ devDefault: 'foo' })).toEqualTypeOf<RequiredValidatorSpec<string>>()
+    expectTypeOf(validator<'foo'>({ devDefault: 'foo' })).toEqualTypeOf<
+      RequiredValidatorSpec<'foo'>
+    >()
+    expectTypeOf(validator({ default: 'foo', devDefault: 'foo' })).toEqualTypeOf<
+      RequiredValidatorSpec<string>
+    >()
+    expectTypeOf(validator<'foo' | 'bar'>({ default: 'foo', devDefault: 'bar' })).toEqualTypeOf<
+      RequiredValidatorSpec<'foo' | 'bar'>
+    >()
+    expectTypeOf(
+      validator<'foo' | 'bar'>({ choices: ['foo', 'bar'], devDefault: 'bar' }),
+    ).toEqualTypeOf<RequiredValidatorSpec<'foo' | 'bar'>>()
+  })
+  test('structured data validator', () => {
+    const validator = makeStructuredValidator(() => ({}))
+    const tt = validator({ default: {} });
+    expectTypeOf(validator()).toEqualTypeOf<RequiredValidatorSpec<any>>()
+    expectTypeOf(validator({ default: {} as any })).toEqualTypeOf<RequiredValidatorSpec<any>>()
+    expectTypeOf(validator({ default: undefined })).toEqualTypeOf<OptionalValidatorSpec<any>>()
+    //@ts-expect-error - Choices not available for structured data
+    validator({ choices: [{ foo: 'bar' }] })
+    expectTypeOf(validator({ devDefault: undefined })).toEqualTypeOf<RequiredValidatorSpec<any>>()
+    expectTypeOf(validator({ devDefault: { foo: 'bar' } })).toEqualTypeOf<
+      RequiredValidatorSpec<{ foo: string }>
+    >()
+    expectTypeOf(validator<{ foo: 'bar' }>()).toEqualTypeOf<RequiredValidatorSpec<{ foo: 'bar' }>>()
+    expectTypeOf(
+      validator({
+        default: {
+          hello: 'world',
+        },
+      }),
+    ).toEqualTypeOf<
+      RequiredValidatorSpec<{
+        hello: string
+      }>
+    >()
+    expectTypeOf(
+      validator<{ hello: 'world' }>({
+        default: {
+          hello: 'world',
+        },
+      }),
+    ).toEqualTypeOf<
+      RequiredValidatorSpec<{
+        hello: 'world'
+      }>
+    >()
+    expectTypeOf(validator<{ hello: string }>()).toEqualTypeOf<
+      RequiredValidatorSpec<{
+        hello: string
+      }>
+    >()
+  })
+})
+
```

Comment:
I like the tests below this line as a general smoke test, but I don't really see the value of the ones above. What's the point in asserting the type of the validator? I think the output value is much more important.

Reply:
To summarize why I did this: 

- All these tests cover, for the most part, the overload complexity of each kind of maker.
- Validator types here are **proxies for the output type**.
- I wanted to test `make*` and `cleanEnv` functions types **in isolation**. It has proven useful when I worked on writing those types, knowing immediately which cases were breaking instead of simply having one `cleanEnv` test failing. 

Hope that helps. If I were to remove all these cases now, we would lose type test coverage. We would have to port a good chunk of these tests to `cleanEnv`, and ideally in a unitary way so that we know where it breaks when it breaks. If you really preferred this, would it be OK if I tackled it in a follow-up PR?

Reply:
I'm good to keep them, can always reassess later if maintenance burden > test value